File: packages/hmssdk_flutter/android/src/main/kotlin/live/hms/hmssdk_flutter/methods/HMSPeerAction.kt
package live.hms.hmssdk_flutter.methods

import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import live.hms.hmssdk_flutter.HMSCommonAction
import live.hms.video.sdk.HMSSDK
import live.hms.video.sdk.models.HMSPeer

class HMSPeerAction {
    companion object{
        // MARK: Peer Actions
        fun peerActions(
            call: MethodCall,
            result: MethodChannel.Result,
            hmssdk: HMSSDK
        ) {
            when (call.method) {
                "change_metadata" -> {
                    changeMetadata(call, result,hmssdk)
                }
                "change_name" -> {
                    changeName(call, result,hmssdk)
                }
                "raise_local_peer_hand"->{
                    raiseLocalPeerHand(result,hmssdk)
                }
                "lower_local_peer_hand"->{
                    lowerLocalPeerHand(result,hmssdk)
                }
                "lower_remote_peer_hand"->{
                    lowerRemotePeerHand(call,result,hmssdk)
                }
                else -> {
                    result.notImplemented()
                }
            }
        }

        private fun changeMetadata(
            call: MethodCall,
            result: MethodChannel.Result,
            hmssdk:HMSSDK
        ) {
            val metadata = call.argument<String?>("metadata")

            metadata?.let {
                hmssdk.changeMetadata(
                    it,
                    hmsActionResultListener = HMSCommonAction.getActionListener(result),
                )
            }
        }

        private fun changeName(
            call: MethodCall,
            result: MethodChannel.Result,
            hmssdk: HMSSDK
        ) {
            val name = call.argument<String?>("name")

            name?.let {
                hmssdk.changeName(
                    name = name,
                    hmsActionResultListener = HMSCommonAction.getActionListener(result),
                )
            }
        }

        private fun lowerLocalPeerHand(
            result: MethodChannel.Result,
            hmssdk: HMSSDK
        ) {
            hmssdk.lowerLocalPeerHand(HMSCommonAction.getActionListener(result))
        }

        private fun raiseLocalPeerHand(
            result: MethodChannel.Result,
            hmssdk: HMSSDK
        ) {
            hmssdk.raiseLocalPeerHand(HMSCommonAction.getActionListener(result))
        }

        private fun lowerRemotePeerHand(
            call: MethodCall,
            result: MethodChannel.Result,
            hmssdk: HMSSDK
        ){
            val peerId = call.argument<String?>("peer_id")
            peerId?.let{
                var forPeer : HMSPeer? = null
                hmssdk.getPeers().forEach{ peer ->
                   if(peer.peerID==peerId) {
                       forPeer = peer
                       return@forEach
                   }
                }
                forPeer?.let {
                    hmssdk.lowerRemotePeerHand(it,HMSCommonAction.getActionListener(result))
                }

            }
        }
    }
}

File: packages/hmssdk_flutter/ios/Classes/Actions/HMSPeerAction.swift
//
//  HMSPeerAction.swift
//  hmssdk_flutter
//
//  Created by Pushpam on 27/09/23.
//

import Foundation
import HMSSDK

class HMSPeerAction {

    static func peerActions(_ call: FlutterMethodCall, _ result: @escaping FlutterResult, _ hmsSDK: HMSSDK?) {
        switch call.method {
        case "change_metadata":
            changeMetadata(call, result, hmsSDK)

        case "change_name":
            changeName(call, result, hmsSDK)

        case "lower_remote_peer_hand":
            lowerRemotePeerHand(call, result, hmsSDK)

        case "raise_local_peer_hand":
            raiseLocalPeerHand(result, hmsSDK)

        case "lower_local_peer_hand":
            lowerLocalPeerHand(result, hmsSDK)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    static private func changeName(_ call: FlutterMethodCall, _ result: @escaping FlutterResult, _ hmsSDK: HMSSDK?) {

        let arguments = call.arguments as![AnyHashable: Any]

        guard let name = arguments["name"] as? String else {
            result(HMSErrorExtension.getError("No name found in \(#function)"))
            return
        }

        hmsSDK?.change(name: name) { _, error in
            if let error = error {
                result(HMSErrorExtension.toDictionary(error))
            } else {
                result(nil)
            }

        }
    }

    static private func changeMetadata(_ call: FlutterMethodCall, _ result: @escaping FlutterResult, _ hmsSDK: HMSSDK?) {

        let arguments = call.arguments as! [AnyHashable: Any]

        guard let metadata = arguments["metadata"] as? String else {
            result(HMSErrorExtension.getError("No metadata found in \(#function)"))
            return
        }

        hmsSDK?.change(metadata: metadata) {_, error in
            if let error = error {
                result(HMSErrorExtension.toDictionary(error))
                return
            } else {
                result(nil)
            }
        }
    }

    static private func lowerLocalPeerHand(_ result: @escaping FlutterResult, _ hmsSDK: HMSSDK?) {
        hmsSDK?.lowerLocalPeerHand {
            _, error in if let error = error {
                result(HMSErrorExtension.toDictionary(error))
            } else {
                result(nil)
            }
        }
    }

    static private func raiseLocalPeerHand(_ result: @escaping FlutterResult, _ hmsSDK: HMSSDK?) {
        hmsSDK?.raiseLocalPeerHand {
            _, error in if let error = error {
                result(HMSErrorExtension.toDictionary(error))
            } else {
                result(nil)
            }
        }
    }

    static private func lowerRemotePeerHand(_ call: FlutterMethodCall, _ result: @escaping FlutterResult, _ hmsSDK: HMSSDK? ) {

        let arguments = call.arguments as? [AnyHashable: Any]

        guard let peerId = arguments?["peer_id"] as? String,
              let forPeer = HMSCommonAction.getPeer(by: peerId, hmsSDK: hmsSDK)
        else {
            HMSErrorLogger.returnArgumentsError("forPeer is null")
            return
        }

        hmsSDK?.lowerRemotePeerHand(forPeer) { _, error in
            if let error = error {
                result(HMSErrorExtension.toDictionary(error))
            } else {
                result(nil)
            }
        }
    }
}
